frontend/src/scenes/data-model/NodeCanvasWithTable.tsx

```diff
@@ -0,0 +1,404 @@
+import clsx from 'clsx'
+import { useEffect, useRef, useState } from 'react'
+
+import GenericNode from './Node'
+import { FixedField, JoinedField, TableFields } from './TableFields'
+import { Edge, Node, NodePosition, NodePositionWithBounds, NodeWithDepth, Position } from './types'
+
+const VERTICAL_SPACING = 150
+const HORIZONTAL_SPACING = 250
+
+const assignDepths = (nodes: Node[]): NodeWithDepth[] => {
+    const nodeMap: { [id: string]: NodeWithDepth } = {}
+
+    // Initialize all nodes with depth -1
+    nodes.forEach((node) => {
+        nodeMap[node.nodeId] = { ...node, depth: -1 }
+    })
+
+    const assignDepthRecursive = (nodeId: string, currentDepth: number): void => {
```

Comment:
We could return this from the backend too. As we traverse the paths, it would be a matter of keeping track of depth.

Reply:
But only if it makes things easier, nothing wrong with this traversal IMO.

Reply:
yep 100%, will add a comment about that. Just did the traversal here for convenience